{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to use this tutorial\n",
    "All hyperlinks are [clickable](IMG1.jpg \"What happened to the curious cat?\"), text in <span title=\"This tutorial was written in Jupyter (iPython notebook) and rendered on Github.\"><b>bold</b></span> can have hovertext with additional information, and all code can be run in a Unix shell (Bash) terminal. \n",
    "\n",
    "# Introduction\n",
    "<span title=\"The process of identifying the order of the individual components in a chain of molecules eg. nucleotides in RNA/DNA or amino acids in proteins.\"><b>Sequencing</b></span> plays an important role in diverse fields including *but not limited to* forensics, evolutionary biology, pharmacology, ecology, and oncology. Nucleic acid sequencing is not exactly a *new* technology; scientists have been using various methods of nucleotide sequencing since the 1970s. [Click here](http://www.sciencedirect.com/science/article/pii/S0888754315300410) to learn more about the history sequencing technology.\n",
    "\n",
    "Have you ever worked with sequencing data before? If your answer was 'Yes', what were you trying to find? How was the data organized? Where did it come from? If your answer was 'No', get ready to try something new!\n",
    "\n",
    "In this tutorial we are going to be working with the model organism *Saccharomyces cerevisiae*. As *S. cerevisiae* is a very well-studied organism, <span title=\"This isn't always a good idea; depending on your research objectives, you may need to consider which criteria were used to decide which features go in the reference annotations and what was discarded.\"><b>we will assume</b></span> that the reference genome & assembly are mostly correct and complete.\n",
    "\n",
    "What kind of organism is *S. cerevisiae*? *S. cerevisiae* is a budding yeast; a <span title=\"It is simple in that yeast is among the easiest organisms to grow in a laboratory setting. It is only biologically simple relative to other multicellular eukaryotes like H. sapiens\"><b>simple</b></span> eukaryote with 16 chromosomes and <span title=\"According to yeastgenome.org, there are 5155 'validated' ORFs. According to the NCBI, there are 6350 genes.\"><b>~6000 genes</b></span>. \n",
    "## Next Generation Sequencing (NGS)\n",
    "Next generation sequencing [also known as  \n",
    "\n",
    "\n",
    "## Methods\n",
    "First you need to download and install the following programs:\n",
    "* [FastQC 0.11.5 for Linux](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.5.zip) or [FastQC 0.11.5 for OSX (Mac)](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.5.dmg)\n",
    "* [Trimmomatic 0.67](http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.36.zip)\n",
    "* [Bowtie2 2.2.9 for Linux](https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.2.9/bowtie2-2.2.9-linux-x86_64.zip/download) or [Bowtie2 2.2.9 for OSX (Mac)](https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.2.9/bowtie2-2.2.9-macos-x86_64.zip/download)\n",
    "* [Cufflinks 2.2.1 for Linux](http://cole-trapnell-lab.github.io/cufflinks/assets/downloads/cufflinks-2.2.1.Linux_x86_64.tar.gz)  or [Cufflinks 2.2.1 for OSX (Mac)](http://cole-trapnell-lab.github.io/cufflinks/assets/downloads/cufflinks-2.2.1.OSX_x86_64.tar.gz)  \n",
    "\n",
    "\n",
    "## Potential problems and pitfalls\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span title=\"\"><b> </b></span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download your reference genome and annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2016-09-05 15:59:17--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.fna.gz\n",
      "           => “GCF_000146045.2_R64_genomic.fna.gz”\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.13, 2607:f220:41e:250::12\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.13|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF_000146045.2_R64 ... done.\n",
      "==> SIZE GCF_000146045.2_R64_genomic.fna.gz ... 3843463\n",
      "==> PASV ... done.    ==> RETR GCF_000146045.2_R64_genomic.fna.gz ... done.\n",
      "Length: 3843463 (3.7M) (unauthoritative)\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  1%  153K 24s\n",
      "    50K .......... .......... .......... .......... ..........  2%  388K 17s\n",
      "   100K .......... .......... .......... .......... ..........  3%  388K 14s\n",
      "   150K .......... .......... .......... .......... ..........  5%  389K 13s\n",
      "   200K .......... .......... .......... .......... ..........  6%  389K 12s\n",
      "   250K .......... .......... .......... .......... ..........  7% 68.4M 10s\n",
      "   300K .......... .......... .......... .......... ..........  9%  391K 9s\n",
      "   350K .......... .......... .......... .......... .......... 10% 49.0M 8s\n",
      "   400K .......... .......... .......... .......... .......... 11% 52.6M 7s\n",
      "   450K .......... .......... .......... .......... .......... 13%  393K 7s\n",
      "   500K .......... .......... .......... .......... .......... 14% 96.2M 6s\n",
      "   550K .......... .......... .......... .......... .......... 15% 61.0M 6s\n",
      "   600K .......... .......... .......... .......... .......... 17%  392K 6s\n",
      "   650K .......... .......... .......... .......... .......... 18% 53.0M 5s\n",
      "   700K .......... .......... .......... .......... .......... 19% 52.7M 5s\n",
      "   750K .......... .......... .......... .......... .......... 21% 78.8M 5s\n",
      "   800K .......... .......... .......... .......... .......... 22% 60.4M 4s\n",
      "   850K .......... .......... .......... .......... .......... 23%  398K 4s\n",
      "   900K .......... .......... .......... .......... .......... 25% 45.1M 4s\n",
      "   950K .......... .......... .......... .......... .......... 26% 56.6M 4s\n",
      "  1000K .......... .......... .......... .......... .......... 27% 54.4M 4s\n",
      "  1050K .......... .......... .......... .......... .......... 29% 55.4M 3s\n",
      "  1100K .......... .......... .......... .......... .......... 30% 81.4M 3s\n",
      "  1150K .......... .......... .......... .......... .......... 31% 56.8M 3s\n",
      "  1200K .......... .......... .......... .......... .......... 33%  404K 3s\n",
      "  1250K .......... .......... .......... .......... .......... 34% 87.5M 3s\n",
      "  1300K .......... .......... .......... .......... .......... 35% 83.1M 3s\n",
      "  1350K .......... .......... .......... .......... .......... 37% 83.4M 3s\n",
      "  1400K .......... .......... .......... .......... .......... 38%  119M 2s\n",
      "  1450K .......... .......... .......... .......... .......... 39% 90.3M 2s\n",
      "  1500K .......... .......... .......... .......... .......... 41% 68.5M 2s\n",
      "  1550K .......... .......... .......... .......... .......... 42% 64.5M 2s\n",
      "  1600K .......... .......... .......... .......... .......... 43% 98.4M 2s\n",
      "  1650K .......... .......... .......... .......... .......... 45% 67.0M 2s\n",
      "  1700K .......... .......... .......... .......... .......... 46%  407K 2s\n",
      "  1750K .......... .......... .......... .......... .......... 47% 36.6M 2s\n",
      "  1800K .......... .......... .......... .......... .......... 49% 73.6M 2s\n",
      "  1850K .......... .......... .......... .......... .......... 50% 98.9M 2s\n",
      "  1900K .......... .......... .......... .......... .......... 51%  159M 1s\n",
      "  1950K .......... .......... .......... .......... .......... 53% 76.5M 1s\n",
      "  2000K .......... .......... .......... .......... .......... 54% 72.9M 1s\n",
      "  2050K .......... .......... .......... .......... .......... 55% 73.8M 1s\n",
      "  2100K .......... .......... .......... .......... .......... 57% 75.3M 1s\n",
      "  2150K .......... .......... .......... .......... .......... 58%  103M 1s\n",
      "  2200K .......... .......... .......... .......... .......... 59%  121M 1s\n",
      "  2250K .......... .......... .......... .......... .......... 61% 85.5M 1s\n",
      "  2300K .......... .......... .......... .......... .......... 62% 83.0M 1s\n",
      "  2350K .......... .......... .......... .......... .......... 63% 75.0M 1s\n",
      "  2400K .......... .......... .......... .......... .......... 65%  414K 1s\n",
      "  2450K .......... .......... .......... .......... .......... 66% 53.8M 1s\n",
      "  2500K .......... .......... .......... .......... .......... 67%  105M 1s\n",
      "  2550K .......... .......... .......... .......... .......... 69% 59.0M 1s\n",
      "  2600K .......... .......... .......... .......... .......... 70% 56.7M 1s\n",
      "  2650K .......... .......... .......... .......... .......... 71% 54.2M 1s\n",
      "  2700K .......... .......... .......... .......... .......... 73% 59.0M 1s\n",
      "  2750K .......... .......... .......... .......... .......... 74% 59.3M 1s\n",
      "  2800K .......... .......... .......... .......... .......... 75% 74.1M 1s\n",
      "  2850K .......... .......... .......... .......... .......... 77% 79.7M 1s\n",
      "  2900K .......... .......... .......... .......... .......... 78% 61.6M 0s\n",
      "  2950K .......... .......... .......... .......... .......... 79% 99.8M 0s\n",
      "  3000K .......... .......... .......... .......... .......... 81%  107M 0s\n",
      "  3050K .......... .......... .......... .......... .......... 82% 71.9M 0s\n",
      "  3100K .......... .......... .......... .......... .......... 83% 89.3M 0s\n",
      "  3150K .......... .......... .......... .......... .......... 85% 90.6M 0s\n",
      "  3200K .......... .......... .......... .......... .......... 86% 74.5M 0s\n",
      "  3250K .......... .......... .......... .......... .......... 87% 95.9M 0s\n",
      "  3300K .......... .......... .......... .......... .......... 89%  119M 0s\n",
      "  3350K .......... .......... .......... .......... .......... 90% 79.4M 0s\n",
      "  3400K .......... .......... .......... .......... .......... 91% 75.8M 0s\n",
      "  3450K .......... .......... .......... .......... .......... 93%  430K 0s\n",
      "  3500K .......... .......... .......... .......... .......... 94% 83.6M 0s\n",
      "  3550K .......... .......... .......... .......... .......... 95% 66.8M 0s\n",
      "  3600K .......... .......... .......... .......... .......... 97% 79.6M 0s\n",
      "  3650K .......... .......... .......... .......... .......... 98%  104M 0s\n",
      "  3700K .......... .......... .......... .......... .......... 99%  111M 0s\n",
      "  3750K ...                                                   100%  110M=1.9s\n",
      "\n",
      "2016-09-05 15:59:22 (1.95 MB/s) - “GCF_000146045.2_R64_genomic.fna.gz” saved [3843463]\n",
      "\n",
      "--2016-09-05 15:59:22--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.gff.gz\n",
      "           => “GCF_000146045.2_R64_genomic.gff.gz”\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 130.14.250.11, 2607:f220:41e:250::13\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.11|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF_000146045.2_R64 ... done.\n",
      "==> SIZE GCF_000146045.2_R64_genomic.gff.gz ... 1424971\n",
      "==> PASV ... done.    ==> RETR GCF_000146045.2_R64_genomic.gff.gz ... done.\n",
      "Length: 1424971 (1.4M) (unauthoritative)\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  3%  167K 8s\n",
      "    50K .......... .......... .......... .......... ..........  7%  199K 7s\n",
      "   100K .......... .......... .......... .......... .......... 10%  398K 6s\n",
      "   150K .......... .......... .......... .......... .......... 14%  399K 5s\n",
      "   200K .......... .......... .......... .......... .......... 17% 59.6M 4s\n",
      "   250K .......... .......... .......... .......... .......... 21%  399K 3s\n",
      "   300K .......... .......... .......... .......... .......... 25% 56.1M 3s\n",
      "   350K .......... .......... .......... .......... .......... 28%  403K 3s\n",
      "   400K .......... .......... .......... .......... .......... 32% 56.4M 2s\n",
      "   450K .......... .......... .......... .......... .......... 35% 79.0M 2s\n",
      "   500K .......... .......... .......... .......... .......... 39% 1.25M 2s\n",
      "   550K .......... .......... .......... .......... .......... 43%  582K 2s\n",
      "   600K .......... .......... .......... .......... .......... 46% 51.3M 1s\n",
      "   650K .......... .......... .......... .......... .......... 50% 55.6M 1s\n",
      "   700K .......... .......... .......... .......... .......... 53% 58.8M 1s\n",
      "   750K .......... .......... .......... .......... .......... 57%  409K 1s\n",
      "   800K .......... .......... .......... .......... .......... 61% 67.8M 1s\n",
      "   850K .......... .......... .......... .......... .......... 64% 81.6M 1s\n",
      "   900K .......... .......... .......... .......... .......... 68% 83.7M 1s\n",
      "   950K .......... .......... .......... .......... .......... 71% 66.3M 1s\n",
      "  1000K .......... .......... .......... .......... .......... 75% 71.7M 0s\n",
      "  1050K .......... .......... .......... .......... .......... 79% 93.3M 0s\n",
      "  1100K .......... .......... .......... .......... .......... 82%  409K 0s\n",
      "  1150K .......... .......... .......... .......... .......... 86% 87.9M 0s\n",
      "  1200K .......... .......... .......... .......... .......... 89%  111M 0s\n",
      "  1250K .......... .......... .......... .......... .......... 93% 88.6M 0s\n",
      "  1300K .......... .......... .......... .......... .......... 97%  121M 0s\n",
      "  1350K .......... .......... .......... .......... .         100%  101M=1.4s\n",
      "\n",
      "2016-09-05 15:59:25 (972 KB/s) - “GCF_000146045.2_R64_genomic.gff.gz” saved [1424971]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Download the reference genome and annotations for Saccharomyces cerevisiae (S288c)\n",
    "wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.fna.gz \n",
    "wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.gff.gz\n",
    "#Unzip the files\n",
    "gunzip GCF_000146045.2_R64_genomic*\n",
    "#Rename the files with names are easier to remember\n",
    "mv GCF_000146045.2_R64_genomic.fna s_cerevisiae.fasta\n",
    "mv GCF_000146045.2_R64_genomic.gff s_cerevisiae.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Due to the time constraints, we're not going to do our analysis on the whole genome; instead, we will work exclusively with chomosome 10 (ChrX). The Fasta file *s_cerevisiae.fasta* contains the sequences of all 16 chromosomes, but we only need chromosome 10. <span title=\"A lazy way to do this is to open the file s_cerevisiae.fasta in a text editor and search for 'NC_001142.9' which is the ID that corresponds to chromosome 10. Delete everything except for chromosome 10 then save the file as 's_cerevisiae_chr10.fasta'.\"><b>How would you make a new file called *s_cerevisiae_chr10.fasta* that only contains chromosome 10?</b></span> Here is a table which will help you figure out which header+sequence to keep:\n",
    "\n",
    "| ID in s_cerevisiae.fasta | Chromosome |\n",
    "|---------------------|------------|\n",
    "| NC_001133.9         | I          |\n",
    "| NC_001134.8         | II         |\n",
    "| NC_001135.5         | III        |\n",
    "| NC_001136.10        | IV         |\n",
    "| NC_001137.3         | V          |\n",
    "| NC_001138.5         | VI         |\n",
    "| NC_001139.9         | VII        |\n",
    "| NC_001140.6         | VIII       |\n",
    "| NC_001141.2         | IX         |\n",
    "| NC_001142.9         | X          |\n",
    "| NC_001143.9         | XI         |\n",
    "| NC_001144.5         | XII        |\n",
    "| NC_001145.3         | XIII       |\n",
    "| NC_001146.8         | XIV        |\n",
    "| NC_001147.6         | XV         |\n",
    "| NC_001148.4         | XVI        |\n",
    "| NC_001224.1         | Mito       |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "### another way to make s_cerevisiae_chr10.fasta\n",
    "head s_cerevisiae.fasta -n 82242 |tail -n 9323 > s_cerevisiae_chr10.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Quality control with *FastQC*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Trimming RNAseq reads with *Trimmomatic*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "export JAVA_HOME=/soft/java\n",
    "export PATH=$JAVA_HOME/bin:$PATH\n",
    ">TruSeq_Adapter_Index_1\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACATCACGATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_2\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGATGTATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_3\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACTTAGGCATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_4\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_5\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACACAGTGATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_6\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGCCAATATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_7\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCAGATCATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_8\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTTGAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_9\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGATCAGATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_10\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACTAGCTTATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_11\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGGCTACATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_12\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCTTGTAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_13\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTCAACAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_14\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACAGTTCCGTATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_15\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACATGTCAGAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_16\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_18\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTCCGCACATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_19\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTGAAACGATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_20\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTGGCCTTATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_21\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGTTTCGGAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_22\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGTACGTAATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_23\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACGAGTGGATATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_25\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACACTGATATATCTCGTATGCCGTCTTCTGCTTG\n",
    ">TruSeq_Adapter_Index_27\n",
    "GATCGGAAGAGCACACGTCTGAACTCCAGTCACATTCCTTTATCTCGTATGCCGTCTTCTGCTTG\n",
    "EOF\n",
    "\n",
    "\n",
    "mkdir trimmomatic\n",
    "java -jar /projects_eg/projects/william/programs/Trimmomatic-0.35/trimmomatic-0.35.jar  PE -phred33 *_read1.fastq.gz *_read2.fastq.gz trimmomatic/$species.read1_paired.fastq trimmomatic/$species.read1_unpaired.fastq trimmomatic/$species.read2_paired.fastq trimmomatic/$species.read2_unpaired.fastq ILLUMINACLIP:$illumina_adapters:2:30:10:2:true LEADING:30 TRAILING:30 SLIDINGWINDOW:4:15 MINLEN:35\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Mapping RNAseq reads to the reference genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "First we will create a new .fastq file for chromosome 10 using our mapped reads from our previous work. (Not for students to repeat, just creating the files here for convenience)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "export PATH=/soft/bio/sequence/samtools-0.1.18/:$PATH\n",
    "samtools view -ub -F 0x40 /projects_eg/projects/william/from_scratch/normal_assemblies/s_cerevisiae/sorted_filtered_s_cerevisiae_and_ERCC.bam NC_001142.9 | samtools bam2fq - > s_cerevisiae_chr10_read1.fastq\n",
    "samtools view -ub -F 0x80 /projects_eg/projects/william/from_scratch/normal_assemblies/s_cerevisiae/sorted_filtered_s_cerevisiae_and_ERCC.bam NC_001142.9 | samtools bam2fq - > s_cerevisiae_chr10_read2.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create an index\n",
    "First we must create a Bowtie2 index using our reference genome (in this case we're just using chromosome 10) so that our reads can be mapped efficiently to the reference sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"s_cerevisiae_chr10.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  s_cerevisiae_chr10.fasta\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 186437\n",
      "Using parameters --bmax 139828 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 139828 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:01\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 124291 (target: 139827)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 6\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 134005\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 134006\n",
      "Getting block 2 of 6\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 134379\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 134380\n",
      "Getting block 3 of 6\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 114741\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 114742\n",
      "Getting block 4 of 6\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 139385\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 139386\n",
      "Getting block 5 of 6\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 111791\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 111792\n",
      "Getting block 6 of 6\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 111445\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 111446\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 231169\n",
      "fchr[G]: 373463\n",
      "fchr[T]: 517338\n",
      "fchr[$]: 745751\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4443180 bytes to primary EBWT file: s_cerevisiae_chr10.1.bt2\n",
      "Wrote 186444 bytes to secondary EBWT file: s_cerevisiae_chr10.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 745751\n",
      "    bwtLen: 745752\n",
      "    sz: 186438\n",
      "    bwtSz: 186438\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 46610\n",
      "    offsSz: 186440\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3885\n",
      "    numLines: 3885\n",
      "    ebwtTotLen: 248640\n",
      "    ebwtTotSz: 248640\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:01\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 186437\n",
      "Using parameters --bmax 139828 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 139828 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 7; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Avg bucket size: 93218 (target: 139827)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 95084\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 95085\n",
      "Getting block 2 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 9\n",
      "Getting block 3 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 150420\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 150421\n",
      "Getting block 4 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 91869\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 91870\n",
      "Getting block 5 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 118174\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 118175\n",
      "Getting block 6 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 90647\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 90648\n",
      "Getting block 7 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 99632\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 99633\n",
      "Getting block 8 of 8\n",
      "  Reserving size (139828) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 99910\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 99911\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 231169\n",
      "fchr[G]: 373463\n",
      "fchr[T]: 517338\n",
      "fchr[$]: 745751\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4443180 bytes to primary EBWT file: s_cerevisiae_chr10.rev.1.bt2\n",
      "Wrote 186444 bytes to secondary EBWT file: s_cerevisiae_chr10.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 745751\n",
      "    bwtLen: 745752\n",
      "    sz: 186438\n",
      "    bwtSz: 186438\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 46610\n",
      "    offsSz: 186440\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 3885\n",
      "    numLines: 3885\n",
      "    ebwtTotLen: 248640\n",
      "    ebwtTotSz: 248640\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "export PATH=/soft/bio/sequence/bowtie2-2.2.4/:$PATH\n",
    "# program    ||   reference genome     || output base filename\n",
    "bowtie2-build  s_cerevisiae_chr10.fasta  s_cerevisiae_chr10\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map the reads with Bowtie2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "export PATH=/soft/bio/sequence/bowtie2-2.2.4/:$PATH\n",
    "export BOWTIE2_INDEXES='.'\n",
    "\n",
    "bowtie2 -x s_cerevisiae_chr10 -p 2 \\\n",
    "-1 ./s_cerevisiae_chr10_read1.fastq \\\n",
    "-2 ./s_cerevisiae_chr10_read2.fastq \\\n",
    "-S 's_cerevisiae_chr10.sam'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "# 5. Assembling the mapped reads into a transcriptome with *Cufflinks*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "##Run Cufflinks2\n",
    "##DESCRIPTION:transcript assembly (Cufflinks2)\n",
    "\n",
    "### What do you want the files created by this script to be called? (short and descriptive)\n",
    "export BASE_NAME_OF_OUTPUT_FILES=\"s_cerevisiae_cufflinks2\"\n",
    "export REFERENCE_GENOME=\"<dir>/s_cerevisiae.fasta\"\n",
    "export REFERENCE_PROTEOME=\"/projects_eg/projects/william/testing_assemblies/all_assemblers/saccharomyces_cerevisiae_chrI_CDS_proteome.fa\"\n",
    "export REFERENCE_ANNOTATION=\"/projects_eg/projects/william/testing_assemblies/all_assemblers/saccharomyces_cerevisiae_chrI_CDS.gff\"\n",
    "export READ_1=\"/projects_eg/projects/william/raw_fastq_normal/s_cerevisiae/trimmomatic/read1_paired.fastq\"\n",
    "export READ_2=\"/projects_eg/projects/william/raw_fastq_normal/s_cerevisiae/trimmomatic/read2_paired.fastq\"\n",
    "export BAM_FILE=\"paired_mapped_SGD_chrI.bam\"\n",
    "\n",
    "\n",
    "#$ -q normal\n",
    "#$ -N cufflink\n",
    "#$ -cwd\n",
    "#$ -e .\n",
    "#$ -o .\n",
    "#$ -pe serial 4\n",
    "\n",
    "##location of programs on cluster (bowtie, bowtie-build,samtools, cufflinks, etc.)\n",
    "export PATH=/soft/bio/sequence/samtools-0.1.18/:$PATH\n",
    "export PATH=/soft/bio/sequence/bedtools-2.17.0/bin/:$PATH\n",
    "export PATH=/soft/bio/sequence/cufflinks-2.2.0/:$PATH\n",
    "export PATH=/soft/bio/sequence/bowtie2-2.2.3/:$PATH\n",
    "\n",
    "export PATH=/soft/bio/sequence/transrate-1.0.0.beta3:$PATH\n",
    "#dependencies\n",
    "export PATH=/soft/bio/sequence/ncbi-blast-2.2.29+/bin:$PATH\n",
    "export PATH=/soft/bio/sequence/SalmonBeta-v0.3.0/bin:$PATH\n",
    "export PATH=/soft/bio/sequence/bam-read-1.0:$PATH\n",
    "export PATH=/soft/bio/sequence/snap_v1.0:$PATH\n",
    "\n",
    "\n",
    "#######################\n",
    "\n",
    "#cufflinks $BAM_FILE -p 4 --frag-bias-correct $REFERENCE_GENOME --multi-read-correct --library-type fr-firststrand -o $BASE_NAME_OF_OUTPUT_FILES \\\n",
    "#--upper-quartile-norm --total-hits-norm --max-mle-iterations 6000 --max-bundle-frags 999999999 --frag-len-std-dev 86 --frag-len-mean 227 \\\n",
    "#--label $BASE_NAME_OF_OUTPUT_FILES --max-bundle-length 4000000 --min-frags-per-transfrag 20 --min-isoform-fraction .05 --min-intron-length 13 --max-intron-length 1001 --trim-3-avgcov-thresh 5 --trim-3-dropoff-frac 0.2 --max-multiread-fraction 0.33 --overlap-radius 1 \\\n",
    "\n",
    "cufflinks $BAM_FILE -p 4 --frag-bias-correct $REFERENCE_GENOME -g $REFERENCE_ANNOTATION --3-overhang-tolerance 120 --multi-read-correct --library-type fr-firststrand -o $BASE_NAME_OF_OUTPUT_FILES \\\n",
    "--upper-quartile-norm --total-hits-norm --max-mle-iterations 6000 --max-bundle-frags 999999999 --frag-len-std-dev 86 --frag-len-mean 227 \\\n",
    "--label $BASE_NAME_OF_OUTPUT_FILES --max-bundle-length 4000000 --min-frags-per-transfrag 20 --min-isoform-fraction .05 --min-intron-length 13 --max-intron-length 1001 --trim-3-avgcov-thresh 5 --trim-3-dropoff-frac 0.2 --max-multiread-fraction 0.33 --overlap-radius 1 \\\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#create FASTA sequences from generated transcripts (for quality assesment with TransRate and for BLASTing)\n",
    "gffread -w $BASE_NAME_OF_OUTPUT_FILES.fa -g $REFERENCE_GENOME  $BASE_NAME_OF_OUTPUT_FILES/transcripts.gtf\n",
    "gffread -w $BASE_NAME_OF_OUTPUT_FILES.guided.fa -g $REFERENCE_GENOME  $BASE_NAME_OF_OUTPUT_FILES.guided/transcripts.gtf\n",
    "\n",
    "\n",
    "#compare with reference genome\n",
    "#cuffcompare -r $REFERENCE_ANNOTATION $BASE_NAME_OF_OUTPUT_FILES/transcripts.gtf -o cmp_$BASE_NAME_OF_OUTPUT_FILES\n",
    "\n",
    "#assess quality of assembly\n",
    "#transrate --assembly $BASE_NAME_OF_OUTPUT_FILES.fa --reference $REFERENCE_PROTEOME --threads 4 -o $BASE_NAME_OF_OUTPUT_FILES -l $READ_1 -i $READ_2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. etc"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
